lib/player/smart_computer.rb

```diff
@@ -0,0 +1,87 @@
+require_relative '../Token'
+require_relative '../board_transformer'
+
+class SmartComputer
+    include Token
+    include BoardTransformer
+    attr_accessor :token
+
+    def move(board:, presenter:)
+        @board = board
+        @transformed_boards = BoardTransformer.transform_with_index(board)
+        @winning_token_count = board.size 
+
+        take_winning || 
+        take_blocking(row_token_count: winning_token_count - 1) || 
+        take_center || 
+        take_empty_corner || 
+        take_row_completing_move(row_token_count: winning_token_count - 2) || 
+        take_blocking(row_token_count: winning_token_count - 2) || 
+        take_random_position
+    end
+
+    private
+    attr_accessor :transformed_boards, :board, :winning_token_count
+
+    def take_winning
+        find_optimal_move(token_to_find: token, target_token_count: winning_token_count - 1)
+    end
+
+    def take_blocking(row_token_count:)
+        find_optimal_move(token_to_find: Token::X, target_token_count: row_token_count)
+    end
+
+    def take_row_completing_move(row_token_count: )
+        find_optimal_move(token_to_find: token, target_token_count: row_token_count) 
+    end
+
+    def find_optimal_move(token_to_find: , target_token_count:)
+        row = find_row_with_token_count(token_to_find: token_to_find, target_token_count: target_token_count)
+        find_empty_position(row) if row
+    end
+
+    def find_row_with_token_count(token_to_find:, target_token_count:)
+        transformed_boards.find do |transformed_board|
+            transformed_board.find do |row|
+                row_token_count = count_tokens(row)
+    
+                has_potential_win = row_token_count[token_to_find] >= target_token_count
+                has_empty_position = row_token_count[Token::EMPTY] >= 1
+    
+                return row if (has_potential_win && has_empty_position)
```

Comment:
First time I've seen it done like this, definitely felt like I got some brain wrinkles sorting out what was going on 🤘